### ISSUE TITLE
[Bug] modeline is not updated when buffer-position is at the right

### ISSUE THREAD

**Issue by geza-herman** [seagle0128/doom-modeline#607]
### Thank you for the bug report

- [X] I am using the latest version of `doom-modeline` related packages.
- [X] I checked [FAQ](https://github.com/seagle0128/doom-modeline#faq).
- [X] You may also try reproduce the issue using clean environment and minimal configurations with
the command `emacs -Q`.

### Bug description

I have a customised doom modeline, I use the segment `buffer-position` at the right. It worked perfectly before, but I think that commit fc25051acca30237b0a8ebff50460f522904bf45 broke this (I'm not exactly sure which commit broke this, but I suspect that it was this one, because ~two weeks ago it worked OK).

The issue is that when I move the cursor up/down, buffer position is not get updated. It only gets updated when the window starts to scroll. I believe this has to do something with pre-formatting the right side of the mode-line string, so emacs no longer knows about that the modeline has to be refreshed, because it doesn't see that it contains `%l` line counter.

This commit caused other problems, "%" characters started to be missing (I had to modify some of my custom segments, so they contain "%%" instead of "%"). Also, I'm not sure whether this commit is the cause, but the string-width calculation is incorrect (causing incorrect right-alignment), because if the right part of the mode-line string contains "%%", it will be counted as two characters, but emacs will display it as a single "%".

### Steps to reproduce

Modify the modeline with this, and witness that the position is not updated:

```elisp
(doom-modeline-def-modeline 'main '(bar) '(buffer-position))
```

### Expected behavior

buffer position should be updated all the time

### OS

Linux

### Emacs Version

29 (dev)

### Emacs Configurations

Happens with vanilla emacs

**Comment by seagle0128** [comment]
It works on my local env. Did you upgrade emacs-dev recently?

**Comment by geza-herman** [comment]
I updated to https://git.savannah.gnu.org/cgit/emacs.git/commit/?h=emacs-29&id=f91bf9df892417a2e4add6d0d77ac5123a579bfc, and the issue still happens. Here's a screenshot. Notice that the cursor is at line 54, yet modeline displays line 19. Also I noticed that the "%" is missing at the end of the "percent of buffer" display, it should be "25%" instead of just "25". 

![image](https://user-images.githubusercontent.com/27892988/216067662-32873dc8-5bf7-4cf6-8c89-a7fe44340635.png)
